zerver/lib/push_notifications.py

```diff
@@ -759,6 +774,51 @@ def handle_remove_push_notification(user_profile_id: int, message_ids: List[int]
         flags=F('flags').bitand(
             ~UserMessage.flags.active_mobile_push_notification))
 
+def encrypt_apns_payload(payload: Dict[str, Any], key: str) -> Dict[str, Any]:
+    encrypted_payload = copy.deepcopy(payload)
+    body = encrypted_payload['alert']['body']
+    encrypted_body, nonce = encrypt_data(key, body)
+    encrypted_payload['alert']['body'] = encrypted_body
+    zulip_data = encrypted_payload['custom']['zulip']
+    zulip_data.update({
+        'nonce': nonce,
+    })
+    return encrypted_payload
+
+def encrypt_gcm_payload(payload: Dict[str, Any], key: str) -> Dict[str, Any]:
+    encrypted_payload = copy.deepcopy(payload)
+    content = encrypted_payload['content']
+    encrypted_content, nonce = encrypt_data(key, content)
+    encrypted_payload['content'] = encrypted_content
+    encrypted_payload['nonce'] = nonce
+    return encrypted_payload
+
+def generate_encrypted_data(devices: List[PushDeviceToken],
+                            payload: Dict[str, Any],
+                            encrypt_payload: Callable[..., Dict[str, Any]]
+                            ) -> EncryptedData:
+    push_data = []
+    for device in devices:
+        key = device.notification_encryption_key
+        encrypted_payload = encrypt_payload(payload, key)
+        push_data.append((device, encrypted_payload))
+
+    return push_data
```

Comment:
This would be a lot simpler as a list comprehension:
```python
  return [(device, encrypt_payload(payload, device.notification_encryption_key))
          for device in devices]
```

Then, further, it'd simplify things to inline that in the two places it's used. More explicit, too, because `encrypt_payload` could be replaced by the specific function to be used in each case.

Reply:
@gnprice can you have a look at the test failure caused due to this change?

We are separating encrypted and unencrypted devices based on whether they have `notification_encryption_key` or not. So in the case of encrypted devices, we are 100% sure that `notification_encryption_key` is non-null but when this is passed as function parameter, `mypy` expects this to be `Optional[str]`. What's the best way to treat this as `str`?